packages/backend/src/core/balances/providers/ArbitrumBalanceProvider.ts

```diff
@@ -0,0 +1,199 @@
+      calls,
+      blockNumber,
+    )
+
+    return balanceQueries.map(({ holder, assetId }, i) => ({
+      assetId,
+      timestamp,
+      holderAddress: holder,
+      chainId: this.chainId,
+      balance: this.decodeForMulticall(balanceResponses[i], assetId) ?? 0n,
+    }))
+  }
+
+  private async fetchUsingPlainCalls(
+    balanceQueries: BalanceQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<BalanceRecord[]> {
+    const rpcCalls = balanceQueries.map((balanceQuery) => {
+      if (this.isNativeCoin(balanceQuery.assetId)) {
+        return this.arbitrumClient.getBalance(balanceQuery.holder, blockNumber)
+      }
+
+      const tokenAddress = getTokenByAssetId(balanceQuery.assetId).address
+
+      if (!tokenAddress) {
+        throw new Error(
+          `Unknown token address ${balanceQuery.assetId.toString()}`,
+        )
+      }
+
+      const { address: to, data } = encodeErc20BalanceQuery(
+        balanceQuery.holder,
+        tokenAddress,
+      )
+
+      return this.arbitrumClient.call({ to, data }, blockNumber)
+    })
+
+    const responses = await Promise.all(rpcCalls)
+
+    return balanceQueries.map(({ holder, assetId }, i) => {
+      const balanceResponse = responses[i]
+
+      const balance = BigNumber.isBigNumber(balanceResponse)
+        ? balanceResponse.toBigInt()
+        : decodeErc20BalanceQuery(balanceResponse)
+
+      return {
+        assetId,
+        timestamp,
+        holderAddress: holder,
+        chainId: this.chainId,
+        balance,
+      }
+    })
+  }
+
+  private encodeForMulticall({
+    assetId,
+    holder,
+  }: BalanceQuery): MulticallRequest {
+    if (this.isNativeCoin(assetId)) {
+      return encodeMulticallCoinBalanceQuery(
+        holder,
+        this.multiCallClient.getAddress(),
+      )
+    }
+
+    const tokenAddress = getTokenByAssetId(assetId).address
+
+    if (!tokenAddress) {
+      throw new Error(`Unknown token address ${assetId.toString()}`)
+    }
+
+    return encodeErc20BalanceQuery(holder, tokenAddress)
+  }
+
+  private isNativeCoin(assetId: AssetId): boolean {
+    return assetId === AssetId.ETH
+  }
+
+  private decodeForMulticall(response: MulticallResponse, assetId: AssetId) {
+    if (!response.success) {
+      return
+    }
+
+    if (this.isNativeCoin(assetId)) {
+      return decodeMulticallCoinBalanceQuery(response.data)
+    }
+
+    return decodeErc20BalanceQuery(response.data)
+  }
+}
+
+function encodeErc20BalanceQuery(
+  holder: EthereumAddress,
+  tokenAddress: EthereumAddress,
+): MulticallRequest {
+  return {
+    address: tokenAddress,
+    data: Bytes.fromHex(
+      erc20Interface.encodeFunctionData('balanceOf', [holder.toString()]),
+    ),
+  }
+}
+
+function decodeErc20BalanceQuery(response: Bytes): bigint {
+  const [value] = erc20Interface.decodeFunctionResult(
+    'balanceOf',
+    response.toString(),
+  )
+
+  return (value as BigNumber).toBigInt()
+}
+
+function encodeMulticallCoinBalanceQuery(
```

Comment:
Kinda feels more explicit, following `ethers.provoder.getBalance(address)`, WYT?
```suggestion
function encodeMulticallGetBalanceQuery(
```

Reply:
Sounds sweet https://github.com/l2beat/l2beat/pull/1645/commits/857e2e6a54be9c1a2eba878e99fd5f144cf6fc4e